kyaml/fn/framework/example_generator_test.go

```diff
@@ -0,0 +1,76 @@
+package framework_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework/command"
+)
+
+// This example is a re-writing of https://github.com/GoogleContainerTools/kpt/issues/2435.
+// We generate with a strong typed object.
+
+func Example_generator() {
+	if err := command.AsMain(framework.ResourceListProcessorFunc(generate)); err != nil {
+		os.Exit(1)
+	}
+}
+
+type Dashboard struct {
+	Revision string `json:"revision"`
+	ID       string `json:"id"`
+}
+
+// generate generates a ConfigMap.
+func generate(rl *framework.ResourceList) error {
+	if rl.FunctionConfig == nil {
+		return fmt.Errorf("unable for find `functionConfig`")
+	}
+	var data map[string]string
+	rl.FunctionConfig.MustGet(&data, "data")
+	db := Dashboard{
+		Revision: data["revision"],
+		ID:       data["id"],
+	}
+	js, err := fetchDashboard(db)
+	if err != nil {
+		return fmt.Errorf("fetch dashboard: %v", err)
+	}
+
+	cm := corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rl.FunctionConfig.GetName(),
+			Namespace: rl.FunctionConfig.GetNamespace(),
+			Labels: map[string]string{
+				"grafana_dashboard": "true",
+			},
+		},
+		Data: map[string]string{
+			fmt.Sprintf("%v.json", rl.FunctionConfig.GetName()): fmt.Sprintf("%q", js),
+		},
+	}
+	return rl.AddObjectToItems(cm, nil, false)
```

Comment:
We want generators to be idempotent. I think the way this is written, it adds Object to items instead overwriting the one that exists, right ? We should probably have a method on `rl.UpsertObject` that add or updates ?